### DISCUSSION TITLE
Add Ability for Explicit Returns: Add The "Return" Keyword

### DISCUSSION THREAD

**Discussion by Mac-Mann** [gleam-lang/gleam#2695] [Ideas & suggestions]
Hello 👋🏻 

As the title suggests, I think having this as an optional keyword in Gleam to explicitly return in Gleam would help add clarity. In additional, providing the ability to explicitly execute and end a functional invocation in a given block provides more options when designing code.

Finally, given the recent uptick in YT videos and folks becoming aware of Gleam, one thing I have noticed is that not having an option for explicit returns is the biggest turn-off to adopting Gleam held by many.

Curious what people think about this idea? I personally don't see much harm in adding this functionality, and it wouldn't break anything in terms of old Gleam code.

**Comment by lpil** [comment]
Hello! Gleam is an entirely expression based language, so there's no early returns or other indirect flow control. This makes a `return` keyword redundant as it can only be in one place, at the end of the block.

For a return keyword to be useful we would need to permit early returns, but this would become extremely confusing alongside Gleam's `use` expressions and other features that depend on Gleam being expression based.

```gleam
pub fn go() {
  let items = [1, 2, 3]
  use <- list.each(items)
  return "Returning early"
}
```

This looks it returns `"Returning early"`, but it returns `Nil` instead

**Reply by Mac-Mann** [reply]
Thanks Louis for explaining this. Given this, I see my understanding of Gleam was lacking. I myself am new to learning about Gleam. That said, since no one had asked this question or made this suggestion before, hopefully this post provides value to folks learning about Gleam as well.

**Reply by lpil** [reply]
💜
